[Refactor] Phaser stuff cleanup

flx-sta opened the pull request:
## What are the changes the user will see?

n/a

## Why am I making these changes?

- One unnecessary src level directory less
- cleaner structure

## What are the changes from a developer perspective?

- move `typings/phaser/index.d.ts` into `@types` and rename it `phaser.d.ts`
- move phaser class prototype extension into `phaser-extension.ts`
    - Include file in `main.ts`
    - Include file in `test/test-utils/testFileInitialization.ts`
- Make `setRealativePosition` return `this` 

## Screenshots/Videos

n/a

## How to test the changes?

- Run the tests
- Run the game

Everything should feel the same

## Checklist

- [x] Otherwise: **I'm using `beta` as my base branch**
- [x] There is no overlap with another PR?
- [x] The PR is self-contained and cannot be split into smaller PRs?
- [x] Have I provided a clear explanation of the changes?
- [x] Have I tested the changes manually?
<!-- We have heavily optimized our test suite, so please actually run the tests :) -->
- [x] Are all unit tests still passing? (`npm run test:silent`)

flx-sta committed:
rename @types/common to ConditionFn.ts

flx-sta committed:
rename typings/phaser/index.d.ts to phaser.d.ts

and move in into @types/

flx-sta committed:
add #app/phaser-extensions.ts

flx-sta committed:
Use phaser-extensions in main & tests

SirzBenjie reviewed (approved):
This is phenomenal. The QoL from `setPositionRelative` now returning `this` is very much appreciated.

DayKev commented on test/test-utils/testFileInitialization.ts:
```diff
@@ -79,24 +80,8 @@ export function initTestFile() {
-    const offsetY = guideObject.height * (-0.5 + (0.5 - guideObject.originY));
-    this.setPosition(guideObject.x + offsetX + x, guideObject.y + offsetY + y);
-  };
+  // await import("#app/phaser-extensions");
```
```suggestion
```
Leftover from testing?

flx-sta replied:
yup. Thanks

DayKev commented on src/@types/phaser.d.ts:
Can we use this to override the type hints of unchanged Phaser functions/etc too?

flx-sta replied:
@DayKev you have an example

flx-sta replied:
Foner

flx-sta replied:
Done. `Tweens.Tween.getValue()` now notes a possible `null` as return value

DayKev commented on src/phaser-extensions.ts:
```diff
@@ -0,0 +1,34 @@
+ * @returns The positioned instance of {@linkcode T}
+ */
+function setPositionRelative<T extends Phaser.GameObjects.GameObject>(
+  guideObject: Phaser.GameObjects.GameObject,
```
Should this be `: T`?

flx-sta replied:
You mean the `guideObjects`? No, this can be something else

Tempo-anon commented on test/test-utils/testFileInitialization.ts:
```diff
@@ -79,24 +80,8 @@ export function initTestFile() {
-    const offsetY = guideObject.height * (-0.5 + (0.5 - guideObject.originY));
-    this.setPosition(guideObject.x + offsetX + x, guideObject.y + offsetY + y);
-  };
+  // await import("#app/phaser-extensions");
```
Is this piece of commented code needed?

flx-sta committed:
Update test/test-utils/testFileInitialization.ts

Co-authored-by: NightKev <34855794+DayKev@users.noreply.github.com>

flx-sta committed:
Merge branch 'beta' into phaser-stuff-cleanup

flx-sta committed:
update phaser.d.ts tsdocs

flx-sta committed:
update phaser.d.ts: update Tweens.getValue()

- add missing `null` for return
- update all `getValue()` to add a `?? 1` as a backup value

innerthunder commented on src/data/mystery-encounters/utils/encounter-pokemon-utils.ts:
```diff
@@ -520,7 +520,7 @@ export function trainerThrowPokeball(
                   onUpdate: (t) => {
                     if (shakeCount && shakeCount < 4) {
-                      const value = t.getValue();
+                      const value = t.getValue() ?? 1;
```
```suggestion
                      const value = t.getValue() ?? 0;
```
Makes more sense imo, since this affects x-position relative to the poke ball's original position

innerthunder commented on src/phases/attempt-capture-phase.ts:
```diff
@@ -137,7 +137,7 @@ export class AttemptCapturePhase extends PokemonPhase {
                 onUpdate: (t) => {
                   if (shakeCount && shakeCount < (isCritical ? 2 : 4)) {
-                    const value = t.getValue();
+                    const value = t.getValue() ?? 1;
```
```suggestion
                    const value = t.getValue() ?? 0;
```
Same here (is this just recycled code?)

DayKev replied:
Yeah MEs copy/pasted the capture code.

flx-sta committed:
Apply suggestions from code review

Co-authored-by: innerthunder <168692175+innerthunder@users.noreply.github.com>

MokaStitcher commented on src/@types/phaser.d.ts:
```diff
@@ -0,0 +1,98 @@
+
+  namespace Tweens {
+    interface Tween {
+      getValue(index?: number): number | null;
```
Where is this coming from?

DayKev replied:
Phaser's return type hint for this function is incorrect, so we're manually fixing it.

MokaStitcher replied:
maybe we should add a comment to explain that, so that when/if they fix it we know we can remove it

MokaStitcher commented on src/phaser-extensions.ts:
```diff
@@ -0,0 +1,34 @@
+
+//#endregion
+
+export default {};
```
what does this do?

flx-sta replied:
Nothing anymore.
When I was experimenting with using `await import("#app/phaser-extensions.ts")` it would make sure that the compiler doesn't complain about the file not being a module, as it wasn't exporting anything.
Removed!

MokaStitcher reviewed (dismissed):
The game won't run on my machine with the following error:
![image](https://github.com/user-attachments/assets/98e3d8bb-723b-4b2d-9b84-1860bca8e856)

otherwise this looks good. This gives a good template on how add `setPositionRelative` to the rex plugin's `BBCodeText` and maybe `InputText` in the future too!

flx-sta committed:
add explanation to Tweens.Tween.getValue

flx-sta committed:
phaser-extensions: remove export, add phaser import

flx-sta committed:
add Phaser.GameObjects.Container.getByType

flx-sta committed:
update phaser.d.ts

- fix some usages of wrongfully assumed properties on a game-object

flx-sta commented:
> The game won't run on my machine with the following error: ![image](https://private-user-images.githubusercontent.com/54149968/437441653-98e3d8bb-723b-4b2d-9b84-1860bca8e856.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDU1OTg1NTMsIm5iZiI6MTc0NTU5ODI1MywicGF0aCI6Ii81NDE0OTk2OC80Mzc0NDE2NTMtOThlM2Q4YmItNzIzYi00YjJkLTliODQtMTg2MGJjYThlODU2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDI1VDE2MjQxM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTE2MzYxMDA4ZTY5NTY1ODQxOTRjZThjZDlmNmE5ZDA0OTY0OGVhMzA1ODc4ZjYxNTNhNThmOTJlNGUwMjNjMjImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.zE2FiFlF1Bmlan7rCr91NNq7SkUWFLLChxLqpcwCjrg)
> 
> otherwise this looks good. This gives a good template on how add `setPositionRelative` to the rex plugin's `BBCodeText` and maybe `InputText` in the future too!

Fixed the missing phaser import.

DayKev commented on src/phaser-extensions.ts:
```diff
@@ -0,0 +1,56 @@
+
+//#region Types
+
+type GuideObject = Phaser.GameObjects.Components.Origin &
+  Phaser.GameObjects.Components.Origin &
+  Phaser.GameObjects.Components.Size &
+  Phaser.GameObjects.Components.Transform;
```
```suggestion
// prettier-ignore
type GuideObject =
  & Phaser.GameObjects.Components.Origin
  & Phaser.GameObjects.Components.Size
  & Phaser.GameObjects.Components.Transform;
```

flx-sta replied:
This would require adjusting prettier somehow. Otherwise it would always be undone

DayKev replied:
Wait, prettier put the `&` at the end of the line? It shouldn't be doing that.

DayKev replied:
Okay I updated the suggestion to add `// prettier-ignore` to this for now... apparently there's a bug in the prettier fork we're using if it's doing this. Since the feature we're using the fork for (which is literally putting stuff like `&` and `||` etc at the start of the line instead of the end) was finally pulled into upstream I'll see if updating to the latest upstream version of prettier fixes this.

DayKev replied:
Okay test successful, upstream prettier has a fixed version of this. Will make a PR to update the deps to upstream Prettier later.

DayKev commented on src/phaser-extensions.ts:
```diff
@@ -0,0 +1,56 @@
+ * @returns The positioned instance of {@linkcode T}
+ */
+function setPositionRelative<T extends Phaser.GameObjects.Components.Transform>(
+  this: T,
```
Using `this` as the name of a param seems dangerous.

flx-sta replied:
No, this is intended.
This help to have proper type-hints for `this`.
it's *auto injected*

flx-sta committed:
Update src/phaser-extensions.ts

Co-authored-by: NightKev <34855794+DayKev@users.noreply.github.com>

flx-sta committed:
add comments for prettier-ignore

MokaStitcher commented on src/@types/phaser.d.ts:
```diff
@@ -0,0 +1,97 @@
+       * @param guideObject - The object to base the position off of
+       * @param x - The relative x position
+       * @param y - The relative y position
+       * @returns The positioned instance of {@linkcode T}
```
Is this doc correct? I know this is copied from the setPositionRelative generic function doc but it seems weird in this context since there's no T I think?

<sub> same with all the others below

flx-sta replied:
Good point. Fixed https://github.com/Despair-Games/poketernity/pull/1066/commits/53a6ca9a494d7c4698fbc16077e443ba594c57d7

MokaStitcher reviewed (dismissed):
much better :p

flx-sta committed:
chore: add nl

flx-sta committed:
Update phaser.d.ts tsdocs

- no longer mentioning `T` inside setPositionRelative @returns, but instead the actual type

MokaStitcher reviewed (dismissed):
should we wait for #1070 so that we can remove the prettier ignore?

flx-sta committed:
Merge branch 'beta' into phaser-stuff-cleanup

flx-sta committed:
remove prettier-ignore from phaser-extensions
